scripts/testing/database-emulator-test.js

```diff
@@ -0,0 +1,139 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const fs = require('fs');
+const path = require('path');
+const request = require('request');
+const tmp = require('tmp');
+
+const { spawn } = require('child-process-promise');
+
+const EMULATOR_LINK =
+  'https://storage.googleapis.com/firebase-preview-drop/emulator/firebase-database-emulator-v3.5.0.jar';
+const EMULATOR_FILENAME = 'database-emulator.jar';
+
+const EMULATOR_PORT = 8088;
+const EMULATOR_NAMESPACE = 'test-emulator';
+
+async function delay(seconds) {
+  return new Promise(resolve => setTimeout(resolve, 1000 * seconds));
+}
+
+async function downloadEmulator() {
+  return new Promise((resolve, reject) => {
+    tmp.dir((err, dir) => {
+      if (err) reject(err);
+      console.log(`Created temporary directory at [${dir}].`);
+      const filepath = path.resolve(dir, EMULATOR_FILENAME);
+      const writeStream = fs.createWriteStream(filepath);
+      console.log(`Downloading emulator from [${EMULATOR_LINK}] ...`);
+      request(EMULATOR_LINK)
+        .pipe(writeStream)
+        .on('finish', () => {
+          resolve(filepath);
+          console.log(`Saved emulator binary file to [${filepath}].`);
+        })
+        .on('error', reject);
+    });
+  });
+}
+
+async function launchEmulator(filepath) {
+  return new Promise((resolve, reject) => {
+    const promise = spawn(
+      'java',
+      ['-jar', path.basename(filepath), '--port', EMULATOR_PORT],
+      {
+        cwd: path.dirname(filepath),
+        stdio: 'inherit'
+      }
+    );
+    promise.catch(reject);
+
+    const timeout = 10; // seconds
+    console.log(`Waiting for emulator to start up ...`);
+    const start = Date.now();
+    setTimeout(function wait() {
+      if (Date.now() - start > timeout * 1000) {
+        reject(`Emulator not ready after ${timeout}s. Exiting ...`);
+      } else {
+        console.log(`Ping emulator at [http://localhost:${EMULATOR_PORT}] ...`);
+        request(`http://localhost:${EMULATOR_PORT}`, (error, response) => {
+          if (error && error.code === 'ECONNREFUSED') {
+            setTimeout(wait, 1000);
+          } else if (response && response.statusCode === 400) {
+            console.log('Emulator has started up successfully!');
+            resolve(promise.childProcess);
+          } else {
+            reject(error);
+          }
+        });
+      }
+    }, 1000);
+  });
+}
+
+async function loadPublicRules() {
+  console.log('Loading rule {".read": true, ".write": true} to emulator ...');
+  return new Promise((resolve, reject) => {
+    request.put(
+      {
+        uri: `http://localhost:${EMULATOR_PORT}/.settings/rules.json?ns=${EMULATOR_NAMESPACE}`,
+        headers: { Authorization: 'Bearer owner' },
+        body: '{ "rules": { ".read": true, ".write": true } }'
+      },
+      (error, response, body) => {
+        if (error) reject(error);
+        console.log(body);
```

Comment:
Is this intentional? Or was it just for debugging?

Reply:
Changed to `console.log(`Done setting public rule to emulator. Response: ${body}.`);`. This is to indicate success set of rules. `body` is {"status":"ok"} in normal cases, which should not be too disruptive.